File: grokkingAlgorithms/javascript/dynamicProgramming/longestSubstring/longestSubstring.js
// Самая длинная общая подстрока

/*
Описание: Допустим, вы открыли сайт dictionary.com. Пользователь вводит слово, а сайт возвращает определение.
Но если пользователь ввел несуществующее слово, нужно предположить, какое слово имелось ввиду.
Пользователь ищет определение fish, но он случайно ввел hish. Такого слова в словаре нет, но зато есть список похожих
слов: fish, vista.
*/

/**
 * Create a matrix
 * @param {number} rows Number of rows
 * @param {number} columns ANumber of columns
 * @returns {Array} matrix
 */
const createMatrix = (rows = 0, columns = 0) => {
    const matrix = [];

    for (let i = 0; i < rows; i++) {
        matrix[i] = Array(columns).fill(0);
    }

    return matrix;
};

/**
 * Find the longest substring
 * @param {string} firstWord First word
 * @param {string} secondWord Second word
 * @returns {string} The longest substring
 */
const longestSubstring = (firstWord = "", secondWord = "") => {
    const matrix = JSON.parse(
        JSON.stringify(createMatrix(firstWord.length, secondWord.length))
    );
    let sizeSequence = 0;
    let indexSequence = 0;

    for (let i = 0; i < firstWord.length; i++) {
        for (let j = 0; j < secondWord.length; j++) {
            if (firstWord[i] === secondWord[j]) {
                matrix[i][j] = (i && j) > 0 ? matrix[i - 1][j - 1] + 1 : 1;

                if (matrix[i][j] >= sizeSequence) {
                    sizeSequence = matrix[i][j];
                    indexSequence = i + 1;
                }
            } else {
                matrix[i][j] = 0;
            }
        }
    }

    return firstWord.slice(indexSequence - sizeSequence, indexSequence);
};

longestSubstring("vista", "hish"); // "is"
longestSubstring("fish", "hish"); // "ish"

File: grokkingAlgorithms/dynamicProgramming/longestSubstring/golang/longest_substring.go
package main

import "fmt"

// Самая длинная общая подстрока
func main() {
	fmt.Println(substring("vista", "hish")) // "is"
	fmt.Println(substring("fish", "hish"))  // "ish"
}

// createMatrix - функция для создания матрицы размера rows на cols
func createMatrix(rows, cols int) [][]int {
	cell := make([][]int, rows)
	for i := range cell {
		cell[i] = make([]int, cols)
	}

	return cell
}

// substring - функция для поиска самой длинной общей подстроки в строках
func substring(a, b string) string {
	lcs := 0                                 // наибольшая общая подстрока
	lastSubIndex := 0                        // индекс последней подстроки
	cell := createMatrix(len(a)+1, len(b)+1) // матрица размером len(a)+1 на len(b)+1

	for i := 1; i <= len(a); i++ {
		for j := 1; j <= len(b); j++ {
			// если символы совпадают, увеличиваем значение ячейки на 1
			if a[i-1] == b[j-1] {
				// если символы не совпадают, выбираем максимальное значение из предыдущих ячеек
				cell[i][j] = cell[i-1][j-1] + 1
				// Если ячейка выше больше, чем ячейка слева, то присваиваем ячейке текущей ячейку выше
				if cell[i][j] > lcs {
					lcs = cell[i][j] // наибольшая общая подстрока
					lastSubIndex = i // индекс последней подстроки
				}
				// Если символы не совпадают, выбираем максимальное значение из предыдущих ячеек
			} else {
				cell[i][j] = 0
			}
		}
	}

	// возвращаем подстроку
	return a[lastSubIndex-lcs : lastSubIndex]
}
